### ISSUE TITLE
Ampersand release management

### ISSUE THREAD

**Issue by Michiel-s** [AmpersandTarski/Ampersand#402]
@hanjoosten I saw that you've released Ampersand v3.5.0 on github. I really love this next step in our maturity. I brings me to the question, what is our release management procedure. 

Things to decide/agree on (and my first suggestions):
* Who decides when to release
 * Joint decision of repository [master pullers team](https://github.com/orgs/AmpersandTarski/teams/masterpullers)
* How often do we want to release. 
 * From my information standardization experience I know: little releases is bad, but too many releases is even worse (e.g. if we release every commit to the master, there's no point in having releases). See timing with XYZ release policy below
* What are the conditions for an X.Y.Z. release.
 * X release concern major changes due to new functionality, changes to existing syntax, or other changes, with **high impact** on end-users. X releases can break backward compatibility and therefore may require Ampersand users to update their scripts (and include folder) accordingly. We aim at max 1 X release every year.
 * Y releases concern major changes due to new/improved functionality, added syntax, refactoring or other changes, with **low impact** on end-users. Y releases are backward compatible and occur when the size of progress justifies a release.
 * Z releases concern bugfixes or refactoring of existing functionality.  We aim to Z release once a month. Depending on the amount of fixes and severity of a certain bug, Z release can occur more often.
* All releases must pass Travis tests (already enforced when merging to master branch)
* All releases must have release notes describing changes (bullet list is ok)
* Please add....

**Comment by hanjoosten** [comment]
I have been looking around what other teams do. For us, I guess we should keep it simple, and use the good stuff invented elsewhere. I think we should adapt to [feature branch workflow](https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow). Later on, we can extend that way of workiing and adapt to [gitflow workflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow). 

The core idea of [feature branch workflow](https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow) is that each change is triggered by its own issue. So for each issue, a specific branch is created. It is developed, and once the developer(s) think it is ready, they do a pull request. Then, other collaborators inspect the code, do some testing. Once there is agreement that the fix/enhancement is good enough, then it can be pulled into the master. 

In [gitflow workflow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow) we do the same thing, but there is also a develop branch. 

Once we adapt to this workflow, doing releases is made more easy.

**Comment by Michiel-s** [comment]
So far you state nothing new to our current process. However, you added a release tag to a certain master commit earlier this morning. This is new and for me reason to open discussion about when and how to assign release tags to certain commits. This is something else than the feature branch workflow.

E.g. are you/we going to add a release tag to every (merge) commit in the master?? I personally don't think that is a good idea. In more general terms, the version numbering (and release management procedure) of Ampersand is unclear to me. See this ticket as an attempt to agree on the procedure.

**Comment by hanjoosten** [comment]
Have a look at [semantic versioning](http://semver.org/) best practises. This works well for anything that has an API. 

I don't know how often we should do a release. I guess this depends on [how much work has been done in a period of time](https://github.com/AmpersandTarski/Ampersand/graphs/contributors).

 The release of this morning was an attempt to just start. I am still struggeling with the release notes. For that reason, we [probably want to annotate commit messages](http://stackoverflow.com/questions/3523534/good-ways-to-manage-a-changelog-using-git), so release notes can be created automatically.

**Comment by hanjoosten** [comment]
I have added a development branch. It contains a releasenotes file. The idea is that from now on, only feature branches will be pushed into the development branch. The master branch is the stable production branch.
